psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -720,192 +768,882 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+                }
+            }
+
+            std::sort(shell_values.begin(), shell_values.end(), ket_compare);
+
+            for (const auto& value : shell_values) {
+                sig_R_for_P[dP].push_back(std::get<0>(value));
+            }
+        }
+
+        // Significant S for P
+        for (int P = Pstart; P < Pstart + nPshell; P++) {
+            int dP = P - Pstart;
+            std::vector<std::tuple<int, double>> shell_values;
+            for (int S = Sstart; S < Sstart + nSshell; S++) {
+                int dS = S - Sstart;
+                double D_max = ints[0]->shell_pair_max_density(P, S);
+                double screen_val = D_max * shell_ceiling_P[dP] * shell_ceiling_S[dS];
+                if (screen_val > linK_ints_cutoff_) {
+                    shell_values.push_back(std::make_tuple(S, screen_val));
+                }
+            }
+
+            std::sort(shell_values.begin(), shell_values.end(), ket_compare);
+
+            for (const auto& value : shell_values) {
+                sig_S_for_P[dP].push_back(std::get<0>(value));
+            }
+
+        }
+
+        // Significant R for Q
+        for (int Q = Qstart; Q < Qstart + nQshell; Q++) {
+            int dQ = Q - Qstart;
+            std::vector<std::tuple<int, double>> shell_values;
+            for (int R = Rstart; R < Rstart + nRshell; R++) {
+                int dR = R - Rstart;
+                double D_max = ints[0]->shell_pair_max_density(Q, R);
+                double screen_val = D_max * shell_ceiling_Q[dQ] * shell_ceiling_R[dR];
+                if (screen_val > linK_ints_cutoff_) {
+                    shell_values.push_back(std::make_tuple(R, screen_val));
+                }
+            }
+
+            std::sort(shell_values.begin(), shell_values.end(), ket_compare);
+
+            for (const auto& value : shell_values) {
+                sig_R_for_Q[dQ].push_back(std::get<0>(value));
+            }
+
+        }
+
+        // Significant S for Q
+        for (int Q = Qstart; Q < Qstart + nQshell; Q++) {
+            int dQ = Q - Qstart;
+            std::vector<std::tuple<int, double>> shell_values;
+            for (int S = Sstart; S < Sstart + nSshell; S++) {
+                int dS = S - Sstart;
+                double D_max = ints[0]->shell_pair_max_density(Q, S);
+                double screen_val = D_max * shell_ceiling_Q[dQ] * shell_ceiling_S[dS];
+                if (screen_val > linK_ints_cutoff_) {
+                    shell_values.push_back(std::make_tuple(S, screen_val));
+                }
+            }
+
+            std::sort(shell_values.begin(), shell_values.end(), ket_compare);
+
+            for (const auto& value : shell_values) {
+                sig_S_for_Q[dQ].push_back(std::get<0>(value));
+            }
+        }
+
+        // Stripeout Lists, used in the "Stripe out" section of the code
+        // Defined after the forthcoming loop
+        std::vector<std::unordered_set<int>> stripeout_R_for_P(nPshell);
```

Comment:
I'm not entirely convinced that std::vector<std::unordered_set<int>> is better than std::vector<std::vector<int>> for these stripeout variables. Each of these stripeout lists' elements are iterated across later on, which will be more efficient with a std::vector compared to a std::unordered_set. However, I don't see where in the algorithm that the O(1) lookup displays its benefits, or where other benefits of the unordered_set come in (e.g., O(1) element find time).

Reply:
I agree, I was originally thinking that if a shell is already in the list, we won't add it twice with this loop design. However, based on the design of the loops, there is no way of that happening.

Reply:
Oops, doing that broke my code. Let me explain why unordered_set is superior. When we add significant Rs and Ss for a significant P or Q, each S is iterated through for each R, so there is a chance that S is already there once we go to the second R shell, and so on. An unordered_set, in this case, is better since looking for an element in a vector to check if it's already there takes O(N) time, whereas it is O(1) in an unordered_set. Both containers have the advantage of O(1) adding time too.